### PR TITLE
LB-1459 Add integration tests for websockets backend

### PR DESCRIPTION
**Pull request by munish0838**
# Problem

LB1459 - https://tickets.metabrainz.org/browse/LB-1459

# Solution
Added websockets container to test docker compose image and tets.sh. Create test_websockets.py to connect to websockets and try to send a valid single json

### CODE REVIEW

**Comment by mayhem** [comment]
I tried this PR locally, but I get:

E           socketio.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=8102): Max retries exceeded with url: /socket.io/?transport=polling&EIO=4&t=1712056391.9411511 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7c5b87d24450>: Failed to establish a new connection: [Errno 111] Connection refused'))

When the test runs. Can you please take another look at it?

**Comment by munish0838** [comment]
thanks @mayhem, will looks into this

**Comment by mayhem** [comment]
Any progress on this? I can't stress how important this PR is -- without a successful PR merged, we won't consider your application further.

File: docker/docker-compose.test.yml [hunk 1]
```diff
@@ -69,3 +69,15 @@ services:
       target: listenbrainz-frontend-dev
     volumes:
       - ../frontend:/code/frontend:z
+
+  websockets:
+    image: web
```
Comment (amCap1712) [inline docker/docker-compose.test.yml:74]: the image here needs to be listenbrainz

File: listenbrainz/tests/integration/test_websockets.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from listenbrainz import db
+from listenbrainz.tests.integration import ListenAPIIntegrationTestCase
+from listenbrainz.webserver.views.api_tools import is_valid_uuid
+import listenbrainz.db.external_service_oauth as db_oauth
+
+class WebSocketTests(ListenAPIIntegrationTestCase):
+    def setUp(self):
+        super(WebSocketTests, self).setUp()
+        self.sio = socketio.Client(logger=True, engineio_logger=True)
+        self.sio.connect('http://localhost:8102',wait_timeout = 20)
```
Comment (amCap1712) [inline listenbrainz/tests/integration/test_websockets.py:22]: url here needs to be http://websockets:8102

File: listenbrainz/tests/integration/test_websockets.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        # Send a listen
+        print(payload)
+        ts = int(time.time())
+        payload['payload'][0]['listened_at'] = ts
+        response = self.send_data(payload)
+        print(response)
+        self.assert200(response)
+        self.assertEqual(response.json['status'], 'ok')
+
+    def send_data(self, payload, user):
```
Comment (amCap1712) [inline listenbrainz/tests/integration/test_websockets.py:39]: You can delete this method and just call the one from base class.

File: requirements_development.txt [hunk 4]
```diff
@@ -5,3 +5,5 @@ pytest==7.1.0
 pytest-cov==3.0.0
 requests-mock==1.9.3
 pytest-subtests==0.8.0
+python-socketio[client]
+python-socketio[asyncio_client]
```
Comment (amCap1712) [inline requirements_development.txt:9]: We only need the first client methinks

File: listenbrainz/tests/integration/test_websockets.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import socketio
+import json
+import time
+
+from listenbrainz.tests.integration import ListenAPIIntegrationTestCase
+
+
+class WebSocketTests(ListenAPIIntegrationTestCase):
```
Comment (amCap1712) [inline listenbrainz/tests/integration/test_websockets.py:8]: I ran this locally and the test now works fine. I suggest you to update the asserts in the listen event to compare some fields in the listen sent and received.

Also, add another test for playing now.

**Comment by amCap1712** [comment]
Hi! The test was hanging up because you had called `self.sio.wait()` in the setup which blocks the thread until disconnect completes so it should be instead be called in the tearDown step.

File: listenbrainz/tests/integration/test_websockets.py (hunk 5 above)
Comment (munish0838) [reply listenbrainz/tests/integration/test_websockets.py:8]: Relieved to know that the tests finally work now. Will work on the assert and playing test now

**Comment by munish0838** [comment]
@amCap1712 Added test for playing now. Both the tests work for me now. Working on adding the comparison of sent and received data since adding those asserts to listen function led to failing of test

**Comment by munish0838** [comment]
@amCap1712, I have updated the tests to include more asset statements for comparing sent and received data

**Comment by munish0838** [comment]
@mayhem Could you please once check if it passes the tests now

**Comment by munish0838** [comment]
> Any progress on this? I can't stress how important this PR is -- without a successful PR merged, we won't consider your application further.
Hey @mayhem , I have added the tests and only pending a review from lucifer now for merging. Please let me know if there is anything else I would need to do too. I hope my application could be considered now.

**Pull request merged**